Adding WAN Lightning support

tv-karthikeya opened the pull request:
Model: lightx2v/Wan2.2-Lightning

Support Wan Unified Transformer on QAIC

tv-karthikeya committed:
Updated WAN and added pytest

Signed-off-by: vtirumal <vtirumal@qti.qualcomm.com>

tv-karthikeya committed:
Refactor Wan pipeline, transformer
  -  Moved blocking to utils from transformer to make it generic
  -  Removed usage of register_modules, components and refactored pipeline, test
  -  Adding Readme, few more constants

Signed-off-by: vtirumal <vtirumal@qti.qualcomm.com>

quic-amitraj commented on pyproject.toml:
```diff
@@ -53,7 +52,12 @@ dependencies = [
 docs = ["Sphinx==7.1.2","sphinx-rtd-theme==2.0.0","myst-parser==3.0.1","sphinx-multiversion"]
 quality = ["black", "ruff", "hf_doc_builder@git+https://github.com/huggingface/doc-builder.git"]
-
+diffusers = [
```
If you will do like this and keeping the wan and flux pipeline importing in init of QEfficient, then you will get error ModuleNotFoundError: No module named 'diffusers'.
There is two solution-

Keep all the packages in the dependencies so this error will resolve.
Remove the imports from the QEfficient init.( Not recommended)

quic-amitraj commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,822 @@
+            height, width, num_frames
+        )
+
+    def calculate_compressed_latent_dimension(self, height, width, num_frames):
```
Do same with this as well. Refer flux pipeline and change accordingly.

quic-amitraj commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,822 @@
+
+    def export(
+        self,
+        height: int = constants.WAN_ONNX_EXPORT_HEIGHT_180P,
```
No need of height and width in export, it make sense to keep in compile. Please refer flux pipeline and make changes based on that.

quic-amitraj commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,822 @@
+            end_time = time.perf_counter()
+            print(f"{module_name} export took {end_time - start_time:.2f} seconds")
+
+    @staticmethod
```
Remove and use self.model

quic-amitraj commented on QEfficient/utils/constants.py:
```diff
@@ -152,6 +152,22 @@ def get_models_dir():
+# Wan Transformer Constants
+WAN_TEXT_EMBED_DIM = 5120
+WAN_PROJECTION_DIM = 6
+WAN_ONNX_EXPORT_BATCH_SIZE = 1
```
Better if you can fetch few values from config so there will not be these many entries here.

tv-karthikeya committed:
Refactor wan
  -  Addressed spec file issue
  -  Adding compile only flag by default
  -  Refactor of common utils

Signed-off-by: vtirumal <vtirumal@qti.qualcomm.com>

tv-karthikeya commented on QEfficient/diffusers/pipelines/configs/wan_config.json:
```diff
@@ -0,0 +1,43 @@
+                    "batch_size": "1",
+                    "num_channels": "16",
+                    "num_frames": "21",
+                    "latent_height": "24",
```
updated

tv-karthikeya commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,822 @@
+        _, self.patch_height, self.patch_width = self.transformer.model.config.patch_size
+
+    @property
+    def do_classifier_free_guidance(self):
```
checked, but cant be used from self.model

tv-karthikeya committed:
Updating Jenkins file

Signed-off-by: vtirumal <vtirumal@qti.qualcomm.com>

quic-amitraj commented on pyproject.toml:
```diff
@@ -40,6 +40,9 @@ dependencies = [
     "fire",
     "py7zr",
     "torchmetrics==1.7.0",
+    "ftfy==6.3.1",
```
Do we have legals of these all packages?

quic-amitraj commented on QEfficient/diffusers/pipelines/pipeline_module.py:
```diff
@@ -479,3 +481,163 @@ def compile(self, specializations: List[Dict], **compiler_options) -> None:
+        Returns:
+            str: Path to the exported ONNX model
+        """
+        if export_kwargs is None:
```
No need of this. Please check flux.

quic-amitraj commented on QEfficient/diffusers/pipelines/modeling_utils.py:
```diff
@@ -0,0 +1,446 @@
+import torch
+
+
+def get_attention_blocking_config():
```
Remove all these from here as it belong to model. create a file named - modeling_utils.py in path QEfficient/diffusers/models and put all these methods there.

quic-amitraj commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,757 @@
+        self.expand_timesteps = self.model.config.expand_timesteps
+
+        # Initialize video processor for frame handling and post-processing
+        self.video_processor = VideoProcessor(vae_scale_factor=self.model.vae.config.scale_factor_spatial)
```
Why still?

quic-amitraj commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,757 @@
+        self.num_frames = num_frames
+
+        # Configure pipeline dimensions and calculate compressed latent parameters
+        self.cl, self.latent_height, self.latent_width, self.latent_frames = calculate_latent_dimensions(
```
Please remove

quic-amitraj commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,757 @@
+
+        # Step 3: Encode input prompts using UMT5 text encoder
+        # TODO: Update UMT5 on QAIC
+        start_encoder_time = time.perf_counter()
```
Pass NA here directly, no need of doing this.

tv-karthikeya commented on tests/diffusers/test_wan.py:
```diff
@@ -0,0 +1,532 @@
+    pipeline.cl, pipeline.latent_height, pipeline.latent_width, pipeline.latent_frames = calculate_latent_dimensions(
+        height,
+        width,
+        num_frames,
```
Added as TODO for now

tv-karthikeya commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,757 @@
+        self.vae_decode = model.vae
+
+        # Store all modules in a dictionary for easy iteration during export/compile
+        self.modules = {"transformer": self.transformer}
```
added here too

tv-karthikeya commented on QEfficient/diffusers/pipelines/wan/pipeline_wan.py:
```diff
@@ -0,0 +1,757 @@
+        return os.path.join(os.path.dirname(__file__), "wan_config.json")
+
+    def compile(
+        self,
```
done

tv-karthikeya committed:
Refactored to address review comments

Signed-off-by: vtirumal <vtirumal@qti.qualcomm.com>

tv-karthikeya committed:
Merge branch 'main' into updated_wan_dit

Signed-off-by: Karthikeya <vtirumal@qti.qualcomm.com>

quic-amitraj reviewed (approved):
LGTM now

tv-karthikeya committed:
Increasing Diffusion test timeout

Signed-off-by: vtirumal <vtirumal@qti.qualcomm.com>

quic-rishinr commented on tests/diffusers/flux_test_config.json:
```diff
@@ -1,33 +1,33 @@
-        "num_single_layers": 2,
-        "use_onnx_subfunctions": false
-    },
+                        "height": 256,
```
are these format issues?
